drivers/raw_exec/driver.go

```diff
@@ -0,0 +1,530 @@
+		Type:             base.PluginTypeDriver,
+		PluginApiVersion: "0.0.1",
+		PluginVersion:    "0.1.0",
+		Name:             pluginName,
+	}
+
+	configSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"enabled": hclspec.NewDefault(
+			hclspec.NewAttr("enabled", "bool", false),
+			hclspec.NewLiteral("false"),
+		),
+		"no_cgroups": hclspec.NewDefault(
+			hclspec.NewAttr("no_cgroups", "bool", false),
+			hclspec.NewLiteral("false"),
+		),
+	})
+
+	taskConfigSpec = hclspec.NewObject(map[string]*hclspec.Spec{
+		"command": hclspec.NewAttr("command", "string", true),
+		"args":    hclspec.NewAttr("command", "list(string)", false),
+	})
+
+	capabilities = &drivers.Capabilities{
+		SendSignals: true,
+		Exec:        true,
+		FSIsolation: drivers.FSIsolationNone,
+	}
+)
+
+// The RawExecDriver is a privileged version of the exec driver. It provides no
+// resource isolation and just fork/execs. The Exec driver should be preferred
+// and this should only be used when explicitly needed.
+type RawExecDriver struct {
+	*utils.Eventer
+	config *Config
+	tasks  *taskStore
+
+	// fingerprintCh is a channel which other funcs can send fingerprints to
+	// that will immediately be sent
+	fingerprintCh chan *drivers.Fingerprint
+
+	stopCh chan struct{}
+
+	logger hclog.Logger
+}
+
+type Config struct {
+	// NoCgroups tracks whether we should use a cgroup to manage the process
+	// tree
+	NoCgroups bool `codec:"no_cgroups"`
+
+	// Enabled is set to true to enable the raw_exec driver
+	Enabled bool `codec:"enabled"`
+}
+
+type TaskConfig struct {
+	Command string   `codec:"command"`
+	Args    []string `codec:"args"`
+}
+
+func NewRawExecDriver(logger hclog.Logger) drivers.DriverPlugin {
+	stopCh := make(chan struct{})
+	return &RawExecDriver{
+		Eventer:       utils.NewEventer(stopCh),
+		config:        &Config{},
+		tasks:         newTaskStore(),
+		fingerprintCh: make(chan *drivers.Fingerprint),
+		stopCh:        stopCh,
+		logger:        logger.Named(pluginName),
+	}
+}
+
+func (r *RawExecDriver) PluginInfo() (*base.PluginInfoResponse, error) {
+	return pluginInfo, nil
+}
+
+func (r *RawExecDriver) ConfigSchema() (*hclspec.Spec, error) {
+	return configSpec, nil
+}
+
+func (r *RawExecDriver) SetConfig(data []byte) error {
+	var config Config
+	if err := base.MsgPackDecode(data, &config); err != nil {
+		return err
+	}
+
+	r.config = &config
+	go r.fingerprintNow()
+	return nil
+}
+
+func (r *RawExecDriver) Shutdown(ctx context.Context) error {
+	close(r.stopCh)
+	return nil
+}
+
+func (r *RawExecDriver) TaskConfigSchema() (*hclspec.Spec, error) {
+	return taskConfigSpec, nil
+}
+
+func (r *RawExecDriver) Capabilities() (*drivers.Capabilities, error) {
+	return capabilities, nil
+}
+
+func (r *RawExecDriver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go r.fingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (r *RawExecDriver) fingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
```

Comment:
I also find this confusing. Can it instead be:

```
func (r *RawExecDriver) fingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
        defer close(r.fingerprintCh)
 	ticker := time.NewTimer(0)
 	for {
 		select {
 		case <-ctx.Done():
 			return
                case <-r.ctx.Done(): // Assuming shutdownCh becomes ctx
 			return
 		case <-ticker.C:
 			ticker.Reset(fingerprintPeriod)
 		        ch <- r.fingerprintNow()
 		}
 	}
 }

 func (r *RawExecDriver) fingerprintNow() *driver.Fingerprint {
 	var health drivers.HealthState
 	var desc string
 	attrs := map[string]string{}
 	if r.config.Enabled {
 		health = drivers.HealthStateHealthy
 		desc = "raw_exec enabled"
 		attrs["driver.raw_exec"] = "1"
 	} else {
 		health = drivers.HealthStateUndetected
 		desc = "raw_exec disabled"
 	}
 	return &drivers.Fingerprint{
 		Attributes:        map[string]string{},
 		Health:            health,
 		HealthDescription: desc,
 	}
 }
```

Reply:
So the idea was to have two modes of operations for fingerprinting. The first was to send fingerprints periodically. The second was to immediatly send a fingerprint when something changed. If SetConfig is only ever called once then this doesn't make sense for this driver. But in others I think we'd want a way to immediately send a fingerprint if say StartTask failed on the docker driver because the docker daemon died.